src/rep.c

```diff
@@ -86,7 +94,14 @@ static SEXP vec_rep_each(SEXP x, SEXP times) {
 
   if (times_size == 1) {
     const int times_ = r_int_get(times, 0);
-    out = vec_rep_each_uniform(x, times_);
+
+    if (times_ == 1) {
+      out = x;
+    } else if (times_ == 0) {
+      out = vec_ptype(x, args_empty);
```

Comment:
@lionel- do you happen to remember if there are any cases where `vec_ptype(x)` would be different from `vec_slice_impl(x, vctrs_shared_empty_int)`? I think the second might be slightly safer here, if there is any chance that they could be different

Reply:
Good catch! We're heading towards namedless prototypes, so it would be more correct to slice.

@mgirlich Could you please change this to slicing and add a unit test for the case of a named vector? The result should be empty and named with an empty character vector.

Reply:
Done. I had a quick glance at the source code of `vec_ptype()` and I guess the only special case is `NULL`

``` r
library(vctrs)

vec_rep(NULL, 0)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
vec_ptype(NULL)
#> NULL

# Before and after the PR NULL produces an error in `vec_rep()` and `vec_rep_each()`
vec_rep(NULL, 0)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
vec_rep_each(NULL, 0)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
```

Do you want to handle this? If so, where do you think is the best place?

Reply:
I think we might be missing a `vec_assert()` in these functions:

```r
vec_rep(NULL, 2)
#> Error: Internal error in `vec_slice_impl()`: Unexpected `NULL`.
```

We should have an error like this:

```r
vec_slice(NULL, 2)
#> Error: Input must be a vector, not NULL.
```

So I handle this with a vector assertion and some unit tests for the `NULL` cases.

Reply:
I also added `vec_assert()` to `vec_unrep()` as the previous error was not very helpful

```r
vctrs::vec_unrep(NULL)
#> Error: All columns in a tibble must be vectors.
#> x Column `key` is NULL.
```